### DISCUSSION TITLE
in plotly-dash.How should I apply the FigureResampler if I use the px.line?

### DISCUSSION THREAD

**Discussion by sc303030** [predict-idlab/plotly-resampler#121] [Q&A] [answered]
In the current project, I'm drawing a graph using plotly-dash px.line.
```python
from plotly_resampler import register_plotly_resampler
register_plotly_resampler(mode='auto')
```
he code above works fine.

I want to use FigureResampler, but there is only go.Figure in the example, so I used the same for the px.line.
```python
fig =  FigureResampler(px.line(df[col_name], x="datetime", y=y_axes, color="target", render_mode="webg1"))
```
But the result I wanted didn't come out like register_plotly_resampler. Is there any way I can use it on px.line?

**Comment by jvdd** [comment, marked as answer]
Hi, I tested similar code to what you provided in your question and it worked as expected (the output of `px.line` which is a `go.Figure` was correctly transformed into a `FigureResampler` object).

Hence, in this response I'll answer two questions that in my belief are relevant for this discussion;

a) *How does `register_plotly_resampler(mode='auto')` work?* 
      => I believe this is necessary to understand how & why your results were satisfactory.
b) *How to use FigureResampler?*
      => This answers the second & main part of your question.


#### A. How does `register_plotly_resampler(mode='auto')` work?
This function results in the use of plotly-resampler under the hood, meaning that plotly figures will be wrapped automatically by the plotly-resampler decorators.  
The `mode` argument determines what decorator should be used. The 'auto' mode uses `FigureWidgetResampler` in IPython based environments (e.g., Jupyter notebooks) and `FigureResampler` in other environments (e.g., running .py file in the terminal).
[More details in the documentation](https://predict-idlab.github.io/plotly-resampler/registering.html#module-plotly_resampler.registering)

So depending on the environment in which you were working, either `FigureResampler` or `FigureWidgetResampler` was used to decorate the plotly figures with (under the hood).

#### B. How to use `FigureResampler`?
You can apply `FigureResampler` in 2 ways;
1. using `register_plotly_resampler` with the mode argument set to 'figure'
2. decorate the figure with the `FigureResampler` class

Note however that for the `FigureResampler` you need to call `.show_dash` to display the visualization (or output the object in a cell via IPython.display)!  
More details: https://github.com/predict-idlab/plotly-resampler#usage

---

PS & TLDR: if you were working in a Jupyter notebook, I would suggest you to use `FigureWidgetResampler` instead of `FigureResampler` to obtain the same output as you had with `register_plotly_resampler(mode='auto')` .
